File: javascript/leetcode/897. Increasing Order Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var increasingBST = function (root) {
  let ans = new TreeNode(0);
  let cur = ans;

  void (function inorder(root) {
    if (!root) return;

    inorder(root.left);
    root.left = null;
    cur.right = root;
    cur = root;
    inorder(root.right);
  })(root);

  return ans.right;
};

File: c++/leetcode/897. Increasing Order Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  TreeNode *cur;

public:
  void inorder(TreeNode *);
  TreeNode *increasingBST(TreeNode *root)
  {
    TreeNode *ans = new TreeNode(0);
    cur = ans;
    inorder(root);
    return ans->right;
  }
};

void Solution::inorder(TreeNode *root)
{
  if (!root)
    return;

  inorder(root->left);
  root->left = nullptr;
  cur->right = root;
  cur = root;
  inorder(root->right);
}